ext/gd/tests/gd_create_xbm_non_castable_user_stream.phpt

```diff
@@ -0,0 +1,83 @@
+--TEST--
+imagecreatefromxbm(): uncastable user stream
+--EXTENSIONS--
+gd
+--SKIPIF--
+<?php
+if (!GD_BUNDLED) die("skip requires bundled GD library\n");
+?>
+--FILE--
+<?php
+
+class DummyStreamWrapper
+{
+    /** @var resource|null */
+    public $context;
+
+    /** @var resource|null */
+    public $handle;
+
+    public function stream_cast(int $castAs)
+    {
+        return false;
+    }
+
+    public function stream_close(): void { }
+
+    public function stream_open(string $path, string $mode, int $options = 0, ?string &$openedPath = null): bool
+    {
+        return true;
+    }
+
+    public function stream_read(int $count)
+    {
+        return false;
+    }
+
+    public function stream_seek(int $offset, int $whence = SEEK_SET): bool
+    {
+        return true;
+    }
+
+    public function stream_set_option(int $option, int $arg1, ?int $arg2): bool
+    {
+        return false;
+    }
+
+    public function stream_stat()
+    {
+        return [];
+    }
+
+    public function stream_tell()
+    {
+        return [];
+    }
+
+    public function stream_truncate(int $newSize): bool
+    {
+        return true;
+    }
+
+    public function stream_write(string $data) { }
+
+
+    public function unlink(string $path): bool
+    {
+        return false;
+    }
+}
+stream_wrapper_register('custom', DummyStreamWrapper::class);
+
+$fp = fopen("custom://myvar", "r+");
+
+/* This tests the underlying _php_image_create_from() C function */
+var_dump(imagecreatefromxbm("custom://myvar"));
+fclose($fp);
+
+echo "Done";
+?>
+--EXPECTF--
+Warning: imagecreatefromxbm(): "custom://myvar" is not a valid XBM file in %s on line %d
```

Comment:
The test [fails on Windows](https://ci.appveyor.com/project/php/php-src/builds/45790577/job/b22crd681nqkpcor#L1703) with a different error message:
````
Warning: imagecreatefromxbm(): Cannot represent a stream of type user-space as a STDIO FILE* in %s on line %d
````
This might be because there is no `fopencookie()` on Windows (and there may not be on other systems), so the failures already occurs at:

https://github.com/php/php-src/blob/ca5f668f7cfa741cdbf3e7f9daa008e803abc923/ext/gd/gd.c#L1567

while the current error message would only be emitted at:

https://github.com/php/php-src/blob/master/ext/gd/gd.c#L1605

If the difference is really due to `fopencookie()`, we might need to augment the zend_test extension to be able to check that, and skip the test if there is support for `fopencookie()` on the machine.

Reply:
Right, let me see what I can do for that

Reply:
It doesn't seem to be that because MacOS fails with
```
 ========DIFF========
001+ Warning: imagecreatefromxbm(): "custom://myvar" is not a valid XBM file in /Users/runner/work/php-src/php-src/ext/gd/tests/gd_create_xbm_non_castable_user_stream_no_fopencookie.php on line 7
001- Warning: imagecreatefromxbm(): Cannot represent a stream of type user-space as a STDIO FILE* in %s on line %d
     bool(false)
     Done
========DONE========
FAIL imagecreatefromxbm(): uncastable user stream without fopencookie [ext/gd/tests/gd_create_xbm_non_castable_user_stream_no_fopencookie.phpt] 
```

Reply:
Hmm, there might be an issue in the `PHP_STREAM_CAST_TRY_HARD` code path, namely when `php_stream_copy_to_stream_ex()` is called:

https://github.com/php/php-src/blob/3b5072f6f619b417614a154b7c2f98e1ea337060/main/streams/cast.c#L273

This fails on Windows, but Windows does not `HAVE_COPY_FILE_RANGE`, so maybe the issue is there?

Reply:
Still not :/ this is really baffling